File: pipeline/kite9-pipeline-common/src/main/java/com/kite9/k9server/command/xml/replace/ReplaceTag.java
package com.kite9.k9server.command.xml.replace;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import org.kite9.diagram.dom.elements.ADLDocument;
import org.w3c.dom.Attr;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.NodeList;

import com.kite9.k9server.adl.holder.pipeline.ADLDom;

/**
 * Replaces some tag names and attributes of tags, but keeps contents unaltered.
 * 
 * @author robmoffat
 *
 */
public class ReplaceTag extends AbstractReplaceCommand<Element, Element> {

	public List<String> keptAttributes = Arrays.asList("id");	// just keep ID by default.
	public List<String> keptTags = Collections.emptyList();
	
	private void checkKeptTags(Element e, Element n, Document d) {
		if (keptTags.contains(e.getTagName())) {
			d.renameNode(n, e.getNamespaceURI(), e.getLocalName());
		}
	}

	private void keepImportantAttributes(Element e, Element n) {
		for (String a : keptAttributes) {
			if (e.hasAttribute(a)) {
				n.setAttribute(a, e.getAttribute(a));
			}
		}
	}

	private void copyAttributes(Element from, Element to, boolean matching) {
		for (int i = 0; i < from.getAttributes().getLength(); i++) {
			Attr item = (Attr) from.getAttributes().item(i);
			if (matching == (keptAttributes.contains(item.getNodeName()))) {
				to.setAttribute(item.getNodeName(), item.getNodeValue());
			}
		}
	}

	private void moveContents(Element from, Element to) {
		NodeList toNodes = to.getChildNodes();
		while (toNodes.getLength() > 0) {
			to.removeChild(toNodes.item(0));
		}
		
		NodeList fromNodes = from.getChildNodes();
		while (fromNodes.getLength() > 0) {
			to.appendChild(fromNodes.item(0));
		}
	}
	
	@Override
	protected void doReplace(ADLDom in, Element e, Element n, Element fromContent) {
		ADLDocument doc = in.getDocument();
		doc.adoptNode(n);
		keepImportantAttributes(e, n);
		e.getParentNode().insertBefore(n, e);
		ensureParentElements(e.getParentNode(), n);
		copyAttributes(e, n, true);
		checkKeptTags(e, n, doc);
		moveContents(e, n);
		e.getParentNode().removeChild(e);
		
		LOG.info("Processed replace tag of "+fragmentId);
	}

	@Override
	protected Mismatch same(Element existing, Element with) {
		Mismatch m1 = checkAttributesSame(existing, with);
		return m1 == null ? checkAttributesSame(with, existing) : m1;
	}

	private Mismatch checkAttributesSame(Element a, Element b) {
		NamedNodeMap nnm = a.getAttributes();
		for (int i = 0; i < nnm.getLength(); i++) {
			Attr aa = (Attr) nnm.item(i);
			
			if ((!keptAttributes.contains(aa.getName())) && (!"xmlns".equals(aa.getName()))) {
				String aValue = aa.getValue();
				String bValue = b.getAttribute(aa.getName());
				if (!aValue.equals(bValue)) {
					return () -> "Attribute" + aa.getName()+" differs: "+aValue+" and "+bValue;
				}
			}	
		}
		return null;
	}
	
	@Override
	protected Element getFromContent(ADLDom adl) {
		Element out = decodeElement(from, adl);
		return out;
	}

	@Override
	protected Element getToContent(ADLDom adl) {
		Element out = decodeElement(to, adl);
		return out;
	}

	@Override
	protected Element getExistingContent(ADLDom in) {
		return findFragmentElement(in.getDocument(), fragmentId);
	}
}

File: pipeline/kite9-pipeline-common/src/main/kotlin/com/kite9/k9server/command/xml/replace/ReplaceTag.kt
package com.kite9.k9server.command.xml.replace

import com.kite9.k9server.adl.holder.pipeline.ADLDom
import com.kite9.k9server.command.Command
import com.kite9.k9server.command.CommandContext
import org.w3c.dom.Attr
import org.w3c.dom.Document
import org.w3c.dom.Element
import java.util.*

/**
 * Replaces some tag names and attributes of tags, but keeps contents unaltered.
 *
 * @author robmoffat
 */
open class ReplaceTag : AbstractReplaceCommand<Element, Element>() {

	var keptAttributes = Arrays.asList("id") // just keep ID by default.
    var keptTags = emptyList<String>()

    private fun checkKeptTags(e: Element, n: Element, d: Document) {
        if (keptTags.contains(e.tagName)) {
            d.renameNode(n, e.namespaceURI, e.localName)
        }
    }

    private fun keepImportantAttributes(e: Element, n: Element) {
        for (a in keptAttributes) {
            if (e.hasAttribute(a)) {
                n.setAttribute(a, e.getAttribute(a))
            }
        }
    }

    private fun copyAttributes(from: Element, to: Element, matching: Boolean) {
        for (i in 0 until from.attributes.length) {
            val item = from.attributes.item(i) as Attr
            if (matching == keptAttributes.contains(item.nodeName)) {
                to.setAttribute(item.nodeName, item.nodeValue)
            }
        }
    }

    private fun moveContents(from: Element, to: Element) {
        val toNodes = to.childNodes
        while (toNodes.length > 0) {
            to.removeChild(toNodes.item(0))
        }
        val fromNodes = from.childNodes
        while (fromNodes.length > 0) {
            to.appendChild(fromNodes.item(0))
        }
    }

    override fun doReplace(d: ADLDom, e: Element, n: Element, fromContent: Element, ctx: CommandContext) {
        val doc: Document = d.document
        doc.adoptNode(n)
        keepImportantAttributes(e, n)
        e.parentNode.insertBefore(n, e)
        ensureParentElements(e.parentNode, n, ctx)
        copyAttributes(e, n, true)
        checkKeptTags(e, n, doc)
        moveContents(e, n)
        e.parentNode.removeChild(e)
        ctx.log("Processed replace tag of $fragmentId")
    }

    protected override fun same(existing: Element, with: Element): Command.Mismatch? {
        val m1: Command.Mismatch? = checkAttributesSame(existing, with)
        return if (m1 == null) checkAttributesSame(with, existing) else m1
    }

    private fun checkAttributesSame(a: Element, b: Element): Command.Mismatch? {
        val nnm = a.attributes
        for (i in 0 until nnm.length) {
            val aa = nnm.item(i) as Attr
            if (!keptAttributes.contains(aa.name) && "xmlns" != aa.name) {
                val aValue = aa.value
                val bValue = b.getAttribute(aa.name)
                if (aValue != bValue) {
                    return Command.Mismatch { "Attribute" + aa.name + " differs: " + aValue + " and " + bValue }
                }
            }
        }
        return null
    }

    override fun getFromContent(adl: ADLDom, ctx: CommandContext): Element {
        return ctx.decodeElement(from, adl)
    }

    override fun getToContent(adl: ADLDom, ctx: CommandContext): Element {
        return ctx.decodeElement(to, adl)
    }

    override fun getExistingContent(d: ADLDom, ctx: CommandContext): Element {
        return findFragmentElement(d.document, fragmentId, ctx)
    }
}